src/neural/network_rr.cc

```diff
@@ -91,5 +93,197 @@ std::unique_ptr<Network> MakeRoundRobinNetwork(
+    for (const auto& plane : input) {
+      hash = HashCat({hash, plane.mask});
+      std::uint32_t tmp;
+      std::memcpy(&tmp, &plane.value, sizeof(float));
+      const std::uint64_t value_hash = tmp;
+      hash = HashCat({hash, value_hash});
+    }
+    return hash;
+  }
+  // Adds a sample to the batch.
+  void AddInput(InputPlanes&& input) override {
+    hashes_.push_back(make_hash(input));
+    requests_.emplace_back();
+    inner_->AddInput(std::move(input));
+  }
+  // Do the computation.
+  void ComputeBlocking() override { inner_->ComputeBlocking(); }
+  // Returns how many times AddInput() was called.
+  int GetBatchSize() const override { return inner_->GetBatchSize(); }
+  float Capture(float value, int index) const {
+    requests_[index].push_back(value);
+    return value;
+  }
+  // Returns Q value of @sample.
+  float GetQVal(int sample) const override {
+    return Capture(inner_->GetQVal(sample), sample);
+  }
+  float GetDVal(int sample) const override {
+    return Capture(inner_->GetDVal(sample), sample);
+  }
+  // Returns P value @move_id of @sample.
+  float GetPVal(int sample, int move_id) const override {
+    return Capture(inner_->GetPVal(sample, move_id), sample);
+  }
+  float GetMVal(int sample) const override {
+    return Capture(inner_->GetMVal(sample), sample);
+  }
+  virtual ~RecordComputation() {
+    Mutex::Lock lock(mutex_);
+    std::fstream output(record_file_, std::ios::app | std::ios_base::binary);
+    for (int i = 0; i < hashes_.size(); i++) {
+      uint64_t value = hashes_[i];
+      output.write(reinterpret_cast<const char*>(&value), sizeof(value));
+      int32_t length = requests_[i].size();
+      output.write(reinterpret_cast<const char*>(&length), sizeof(length));
+      for (int j = 0; j < length; j++) {
+        float recorded = requests_[i][j];
+        output.write(reinterpret_cast<const char*>(&recorded),
+                     sizeof(recorded));
+      }
+    }
+  }
+  std::unique_ptr<NetworkComputation> inner_;
+  std::string record_file_;
+  std::vector<uint64_t> hashes_;
+  mutable std::vector<std::vector<float>> requests_;
+  static Mutex mutex_;
+};
+
+Mutex RecordComputation::mutex_;
+
+class ReplayComputation : public NetworkComputation {
+ public:
+  ReplayComputation(std::unordered_map<uint64_t, std::vector<float>>* lookup)
+      : lookup_(lookup) {}
+  // Adds a sample to the batch.
+  void AddInput(InputPlanes&& input) override {
+    hashes_.push_back(RecordComputation::make_hash(input));
+    replay_counter_.push_back(0);
+  }
+  // Do the computation.
+  void ComputeBlocking() override {}
+  // Returns how many times AddInput() was called.
+  int GetBatchSize() const override { return hashes_.size(); }
+  float Replay(int index) const {
+    const auto& entry_ptr = lookup_->find(hashes_[index]);
+    if (entry_ptr == lookup_->end()) {
+      return 0.0f;
+    }
+    const auto& entry = entry_ptr->second;
+    int counter = replay_counter_[index];
```

Comment:
Make `unsigned` to fix warning.

Reply:
Made replay_counter_ size_t and changed.